### PR TITLE
Feature/sandbox

### PR DESCRIPTION
**Pull request by sapieneptus**
## Motivation
- Installing ruby is a PITA, especially for first time users. 
- Users sometimes write tests that work on rubies not present in XTC, and thus they work locally but not in the cloud

## The Idea

Similar to [calabash.app] (https://github.com/calabash/calabash.app), the idea is you can invoke a sandbox environment for calabash that just works. 

## Workflow

```shell
$ ./install-osx.sh
Preparing Ruby...
######################################################################## 100.0%
Preparing sandbox...
Installing gems, this may take a little while...

# does a gem install and a bundle install.

Done!
Execute 'calabash-sandbox' to get started! 

$ calabash-sandbox
This terminal is now ready to use with Calabash.
To exit, type 'exit'.
[calabash-sandbox] $ calabash-android gen  #just works
[calabash-sandbox] $ calabash-ios version  #just works
0.16.4
[calabash-sandbox] $ exit
$ # back to normal
```

@jonstoneman Would be great to do a windows port
@krukow @jmoody I imagine we don't need a linux port?

- [ ] Stand up Jenkins cucumbers
- [ ] Stand up Jenkins submit to test-cloud
- [ ] Check the sandbox on a clean machine (no Xcode or development tools installed)

### CODE REVIEW

**Comment by jmoody** [comment]
> Users sometimes write tests that work on rubies not present in XTC, and thus they work locally but not in the cloud

This PR does not solve this problem, right?

**Comment by sapieneptus** [comment]
@jmoody The idea is this PR installs Ruby 2.0.0, which is currently the one used in XTC. So when you're in the sandbox, if your tests work, they ought to work on XTC as well. 

At some point we'll want to use Ruby 2.1.5 instead (or maybe we should do that now as a precaution?)

**Comment by krukow** [comment]
@sapieneptus we're switching to 2.1.5 very soon so you should just jump to that.

**Comment by jmoody** [comment]
@krukow I checked with @olefriis last week.  I believe XTC is running 2.1.5 already?

**Comment by jmoody** [comment]
@sapieneptus 

```
# Always available; even on Linux.
$ uname -s
Darwin
```

**Comment by jmoody** [comment]
@sapieneptus 

It would be nice to do the following in Jenkins:

* run some cucumbers
* submit a job to the XTC

File: calabash-sandbox [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
```
Comment (jmoody) [inline calabash-sandbox]: This should be:

```
#!/usr/bin/env bash
```

to respect the users installed bash.

**Comment by olefriis** [comment]
@jmoody Regarding Ruby version, I think I gave you the wrong information! I simply logged into an execution host and saw that it had Ruby 2.1.5 installed. However, from stand-ups the last couple of days I've been told that this is _not_ the Ruby used in the Cucumber sandbox.

I still don't know which exact version is being used in the sandbox, but @krukow and/or @chvitved probably know...

**Comment by sapieneptus** [comment]
It is now installing 2.1.5-p273, which should be the same as XTC soon, per [this] (https://github.com/xamarin/documentation/issues/987). 

@jmoody I've made several changes, would you mind re-reviewing?

I've tested this on a brand new user on ElCap and it seems to work - however, I had to take the approach of advising the user to run `./calabash-sandbox` in the `pwd` if unable to install to `/usr/local/bin`, which is likely for a brand new setup. However, it still works exactly the same whether invoked locally or globally. 

It is also now downloading the Gems as a zip file, which is a little slow but much more consistent behavior. 

@krukow I'm not sure how many people will download this, but is it generally safe to keep all of the zip dependencies in S3? Or is there a chance we should use something like cloudfront/glacier? If you or @jmoody happen to know the download stats, that might be a good indicator...

File: install-osx.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    exit 0
+  fi
+fi
+
+mkdir -p "$GEM_HOME"
+mkdir -p "$HOME/.calabash/sandbox/Rubies"
+
+#Download Ruby
+echo "Preparing Ruby ${CALABASH_RUBY_VERSION}..."
+curl -o "${CALABASH_RUBY_VERSION}.zip" --progress-bar https://s3-eu-west-1.amazonaws.com/calabash-files/2.1.5-p273.zip
```
Comment (jonstoneman) [inline install-osx.sh]: That zip contains a `__MACOSX` folder.

Comment (jmoody) [inline install-osx.sh]: +1 for tar.gz

**Comment by jmoody** [comment]
@sapieneptus Feel free to rebase.  In general, I think rebasing a public branch is fine as long as you notify everyone at the top of the PR.

e.g.

**REBASED: ** Fri Nov 14:10 CET

**Comment by jmoody** [comment]
@sapieneptus I will review.  Do you mind if just make changes directly?  I noticed some trivial issue the last time I took a look.

> It is also now downloading the Gems as a zip file, which is a little slow but much more consistent behavior.

Two things will help us here:

1. I believe we can remove at least 3 gem dependencies.
2. The biggest parts of the archive are the Calabash iOS libraries.  Once we have transitioned to calabash-tool, the bundle will become much smaller.

**Comment by jmoody** [comment]
Jenkins test this please

**Comment by jmoody** [comment]
@sapieneptus 

I am working on some integration tests.  I am seeing this output when running cucumber:

```
$ APP=calabash-ios-server/Products/test-target/app-cal/LPTestTarget.app cucumber
Ignoring unf_ext-0.0.7.1 because its extensions are not built.  Try: gem pristine unf_ext --version 0.0.7.1
Ignoring gherkin-2.12.2 because its extensions are not built.  Try: gem pristine gherkin --version 2.12.2
Ignoring gherkin-2.12.2 because its extensions are not built.  Try: gem pristine gherkin --version 2.12.2
```

The cucumber job runs.

**Comment by jmoody** [comment]
I am convinced.  The Jenkins tests actually runs some Calabash iOS cucumbers against an iOS Simulator.

http://calabash-ci.macminicolo.net:8080/job/Calabash-install-PR/24/console

**Pull request merged**